lib/token-manager.js

```diff
@@ -345,37 +349,92 @@ TokenManager.prototype = {
-				// In these cases, we can attempt to retry the grant request with a new exp claim calculated frem the
-				// Date header sent by the server
-				if (isJWTAuthErrorRetryable(err)) {
-
-					var serverTime = Math.floor(Date.parse(err.response.headers.date) / 1000);
-					claims.exp = serverTime + this.config.appAuth.expirationTime;
-					jwtOptions.jwtid = uuid.v4();
+			.catch(err => this.retryJWTGrant(claims, jwtOptions, keyParams, params, options, err, 0));
+	},
 
-					try {
-						params.assertion = jwt.sign(claims, keyParams, jwtOptions);
-					} catch (jwtErr) {
-						throw jwtErr;
+	/**
+	 * Attempt a retry if possible and create a new JTI claim. If the request hasn't exceeded it's maximum number of retries,
+	 * re-execute the request (after the retry interval). Otherwise, propagate a new error.
+	 *
+	 * @param {Object} claims - JWT claims object
+	 * @param {Object} [jwtOptions] - JWT options for the signature
+	 * @param {Object} keyParams - Key JWT parameters object that contains the private key and the passphrase
+	 * @param {Object} params - Should contain all params expected by Box OAuth2 token endpoint
+	 * @param {TokenRequestOptions} [options] - Sets optional behavior for the token grant
+	 * @param {Error} error - Error from the previous JWT request
+	 * @param {int} numRetries - Number of retries attempted
+	 * @returns {Promise<TokenInfo>} Promise resolving to the token info
+	 */
+	// eslint-disable-next-line max-params
+	retryJWTGrant(claims, jwtOptions, keyParams, params, options, error, numRetries) {
+		if (numRetries < this.config.numMaxRetries && isJWTAuthErrorRetryable(error)) {
+			var retryTimeout;
+			numRetries += 1;
+			// If the retry strategy is defined, then use it to determine the time (in ms) until the next retry or to
+			// propagate an error to the user.
+			if (this.config.retryStrategy) {
+				// Get the total elapsed time so far since the request was executed
+				var totalElapsedTime = process.hrtime(asyncRequestTimer);
+				var totalElapsedTimeMS = (totalElapsedTime[0] * 1000) + (totalElapsedTime[1] / 1000000);
+				var retryOptions = {
+					error,
+					numRetryAttempts: numRetries,
+					numMaxRetries: this.config.numMaxRetries,
+					retryIntervalMS: this.config.retryIntervalMS,
+					totalElapsedTimeMS
+				};
+
+				retryTimeout = this.config.retryStrategy(retryOptions);
+
+				// If the retry strategy doesn't return a number/time in ms, then propagate the response error to the user.
+				// However, if the retry strategy returns its own error, this will be propagated to the user instead.
+				if (typeof retryTimeout !== 'number') {
+					if (retryTimeout instanceof Error) {
+						error = retryTimeout;
 					}
-
-					return this.getTokens(params, options);
+					throw error;
 				}
+			} else if (error.response.headers.hasOwnProperty('retry-after')) {
+				retryTimeout = error.response.headers['retry-after'] * 1000;
```

Comment:
Should you add the retry-after check here, as well?
https://github.com/box/box-node-sdk/blob/master/lib/api-request.js#L321